### ISSUE TITLE
Keyboard interaction issues with dynamics popup

### ISSUE THREAD

**Issue by avvvvve** [musescore/MuseScore#26249]
### Issue type

UX/Interaction bug (incorrect behaviour)

### Description with steps to reproduce

Follow-up from #26189. There are 3 parts to this issue:

-----

### 1. Shift interaction

[Original comment](https://github.com/musescore/MuseScore/pull/26189#issuecomment-2608736895)

One problem though is that the shift-to-move-beat-position interaction only seems to work if you select the dynamic with the mouse. It doesn't work if you select the dynamic via keyboard navigation or after having just inputted it via either the popup or the input text field. 

https://github.com/user-attachments/assets/4aa46619-dd14-4de6-aa62-2b2a67fffbd7

Also doesn't work if you add the dynamic from the palette:

https://github.com/user-attachments/assets/65647fd3-12f8-4c5e-8013-9a2ede12180a

-----

### 2. Alt + Left/Right arrow interaction

[Original comment](https://github.com/musescore/MuseScore/pull/26189#issuecomment-2613645556)

When you click on a dynamic to select it, Alt/Option + Left/Right arrow will no longer move the selection to the next/previous element. Instead, it nudges it to the left/right as if just an arrow with no modifier was pressed.

- The same happens even if you get the popup to disappear by some means and then try to Alt + Arrow, i.e....
     - Scrolling the score to close the popup
     - Move focus to the popup with the keyboard, change the dynamic, press esc to close the popup leaving the dynamic selected (shown in the video at 0:11)

https://github.com/user-attachments/assets/2e1d1a19-aaaa-43db-a8f1-fe8edf245f3e

-----

### 3. Shortcuts should still work while focus is in popup

[Original comment](https://github.com/musescore/MuseScore/pull/26189#issuecomment-2615693781)

When you tab to move focus into the popup, some shortcuts work (i.e. voice assignment) but others do nothing. These shortcuts should be possible without having to press esc first:

- Alt + Left/Right arrows to select previous/next element (popup should close)
- Alt + Up/Down arrows to select other elements at that beat position (popup should close)
- Ctrl + C to copy (popup should remain open)
- Ctrl + X to cut (popup should close)

### Supporting files, videos and screenshots

See above

### What is the latest version of MuseScore Studio where this issue is present?

https://github.com/musescore/MuseScore/pull/26189

### Regression

No.

### Operating system

macOS 15

### Checklist

- [x] This report follows the [guidelines](https://github.com/musescore/MuseScore/wiki/Reporting-bugs-and-issues) for reporting bugs and issues
- [x] I have verified that this issue has not been logged before, by searching the [issue tracker](https://github.com/musescore/MuseScore/issues) for similar issues
- [x] I have attached all requested files and information to this report
- [x] I have attempted to identify the root problem as concisely as possible, and have used minimal reproducible examples where possible

**Comment by bkunda** [comment]
Just checking through these issues in the latest nightly:

1. `Shift` now seems to work if adding from the palettes. As for the other interactions, while they fail the first time, they work after deselecting and reselecting the object. This is a P2. 
2. This is indeed an accessibility problem, and makes navigating between score elements more tedious. P1.
3.  I'd say these are also less urgent.

More of a problem is that I cannot seem to TAB to focus on the popup if I've entered the dynamic using the Ctrl+D shortcut:

https://github.com/user-attachments/assets/f06bd3bd-fa9e-4143-91b8-e91f66eca982

In sum, I'd fix this plus point 2 above first, then reclassify the remainder of work on this issue as a P2.

**Comment by MarcSabatella** [comment]
I missed this when looking for previous reports regarding point 2 - so my #26718 is a duplicate.  Definitely would be great to have fixed for 4.5 if at all possible!

Regarding that last point about Tab not moving into the popup - that would be nice to have indeed.  But from an accessibility standpoint, the ability to type the dynamics directly is a huge win already and what I think most blind users would choose anyhow.  So I doubt most blind users would consider this to be anywhere near as critical as the Alt+Left/Right behavior.

**Comment by avvvvve** [comment]
re: Cmd + D @Tantacrul :
@cbjeukendrup mentioned somewhere (can't find it now) that the keyboard shortcut not working will not be an issue for users once we release. You can fix it in your own build by going to Preferences > Shortcuts > find 'Add dynamic' and click 'Reset to default'

As just discussed in standup, Casper is going to take a look at everything else in this issue. So if you'd like me to split it up, I can @cbjeukendrup—your call!

**Comment by Tantacrul** [comment]
OK, all good. I noticed that this was on the MS4.5 board after my comment. Sounds like all is going according to plan. Thanks @avvvvve

**Comment by MarcSabatella** [comment]
Regarding the shortcut:

What usually happens is that new installations will get new shortcuts by default, and so will updates to older installations if the user had *not* previously customized their shortcuts.  But if the user had previously updated shortcuts, then as far as I know the custom definition file is used as is, so any new shortcuts get missed.  In a few cases we've special-cased this to merge in new shortcuts if they don't conflict with existing ones.  I think there was one case where we actually forced the new shortcut to take effect even if the user *had* previously assigned it to something else.  I can't recall if there is is now a generally-useful framework for this so we could easily specify the behavior for the dynamics shortcut without more special-case hacks.

BTW, beta and nightly build use different settings (release vs development), so it's entirely possible that you'd see the new shortcut on one installation but not the other depending on whether you had previously customized shortcuts in the release versus other development builds.  So that could explain why the shortcut works by default in the beta but not nightly for some systems.

**Comment by Tantacrul** [comment]
@MarcSabatella ah, yes. Thanks for that. I had forgotten.

**Comment by cbjeukendrup** [comment]
It's slightly different than that. In early nightly builds, the action code was `dynamics`, and Ctrl+D was bound to that action code. Then I changed the action code to `add-dynamic`, and reassigned Ctrl+D to that code in the default shortcuts file. Normally, shortcuts in the default shortcuts file that are not in the user's shortcut file are copied to the user's shortcuts file. But in this case, that does not happen, because the Ctrl+D shortcut is already in use in the user's shortcuts file for a different action code, namely the old `dynamics` code. So even though no action with the code `dynamics` exists anymore, the fact that the user had Ctrl+D assigned to that code prevents that shortcut from being automatically migrated to the new action code. I have a simple patch to fix that, and I could create a PR for that. But it won't affect end users anyway: they won't have this old `dynamics` code in their shortcuts file (well, if they used those old nightly builds, they might have it in their "development" configuration shortcuts file, but they won't have it in the shortcuts file that's used by the stable release).
